cpmpy/solvers/TEMPLATE.py

```diff
@@ -0,0 +1,283 @@
+            self += cpm_model.constraints
+
+            # post objective
+            if cpm_model.objective is not None:
+                if cpm_model.objective_max:
+                    self.maximize(cpm_model.objective)
+                else:
+                    self.minimize(cpm_model.objective)
+
+
+    def solve(self, time_limit=None, **kwargs):
+        """
+            Call the TEMPLATE solver
+
+            Arguments:
+            - time_limit:  maximum solve time in seconds (float, optional)
+            - kwargs:      any keyword argument, sets parameters of solver object
+
+            Arguments that correspond to solver parameters:
+            <Please document key solver arguments that the user might wish to change
+             for example: log_output=True, var_ordering=3, num_cores=8, ...>
+            <Add link to documentation of all solver parameters>
+        """
+
+        if time_limit is not None:
+            raise NotImplementedError("TEMPLATE: TODO, implement time_limit")
+
+        # call the solver, with parameters
+        my_status = self.TEMPLATE_solver.solve(**kwargs)
+
+        # new status, translate runtime
+        self.cpm_status = SolverStatus(self.name)
+        self.cpm_status.runtime = self.TEMPLATE_solver.time()
+
+        # translate exit status
+        if my_status is True:
+            self.cpm_status.exitstatus = ExitStatus.FEASIBLE
+        elif my_status is False:
+            self.cpm_status.exitstatus = ExitStatus.UNSATISFIABLE
+        elif my_status is None:
+            # can happen when timeout is reached...
+            self.cpm_status.exitstatus = ExitStatus.UNKNOWN
+        else: # another?
+            raise NotImplementedError(my_status) # a new status type was introduced, please report on github
+
+        # True/False depending on self.cpm_status
+        has_sol = self._solve_return()
+
+        # translate solution values (of user vars only)
+        if has_sol:
+            # fill in variable values
+            for cpm_var in self.user_vars:
+                sol_var = self.solver_var(cpm_var)
+                cpm_var._value = None # if not in solution
+                #cpm_var._value = self.TEMPLATEpy.value(sol_var)
+                raise NotImplementedError("TEMPLATE: back-translating the solution values")
+
+        return has_sol
+
+
+    def solver_var(self, cpm_var):
+        """
+            Creates solver variable for cpmpy variable
+            or returns from cache if previously created
+        """
+        # TODO: add `solver_vars(self, cpm_vars)` to SolverInterface class
+
+        if is_num(cpm_var):
+            return cpm_var
+
+        # special case, negative-bool-view
+        # work directly on var inside the view
+        if isinstance(cpm_var, NegBoolView):
+            return TEMPLATEpy.negate(self.solver_var(cpm_var._bv))
+
+        # create if it does not exit
+        if not cpm_var in self.varmap:
```

Comment:
The create if does not exist part:
should we be able to do something like `self += [cpm_var]`?
I.e. posting a constraint with only one variable? It would certainly make the code very simple and could even mean this function can be put in the superclass. At the moment, I don't think we support it. But also not sure if we want something like this...